server/util/rexec/rexec.go

```diff
@@ -0,0 +1,452 @@
+	}
+	res := c.response.GetResult()
+	var stdout, stderr bytes.Buffer
+	eg, egctx := errgroup.WithContext(ctx)
+	if res.GetStdoutDigest() != nil {
+		eg.Go(func() error {
+			ctx := egctx
+			rn := digest.NewResourceName(res.GetStdoutDigest(), c.InstanceName, rspb.CacheType_CAS, c.digestFunction())
+			return cachetools.GetBlob(ctx, c.env.GetByteStreamClient(), rn, &stdout)
+		})
+	}
+	if res.GetStderrDigest() != nil {
+		eg.Go(func() error {
+			ctx := egctx
+			rn := digest.NewResourceName(res.GetStderrDigest(), c.InstanceName, rspb.CacheType_CAS, c.digestFunction())
+			return cachetools.GetBlob(ctx, c.env.GetByteStreamClient(), rn, &stderr)
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+	c.commandResult = &interfaces.CommandResult{
+		ExitCode: int(res.GetExitCode()),
+		Stdout:   stdout.Bytes(),
+		Stderr:   stderr.Bytes(),
+	}
+	return c.commandResult, nil
+}
+
+func (c *Cmd) digestFunction() repb.DigestFunction_Value {
+	if c.DigestFunction != 0 {
+		return c.DigestFunction
+	}
+	return defaultDigestFunction
+}
+
+// RetryingStream implements a reliable operation stream.
+//
+// It keeps track of the operation name internally, and provides a Recv() func
+// which re-establishes the stream transparently if the operation name has been
+// established.
+type RetryingStream struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	client repb.ExecutionClient
+	stream repb.Execution_ExecuteClient
+	name   string
+}
+
+func NewRetryingStream(ctx context.Context, client repb.ExecutionClient, stream repb.Execution_ExecuteClient, name string) *RetryingStream {
+	ctx, cancel := context.WithCancel(ctx)
+	return &RetryingStream{
+		ctx:    ctx,
+		cancel: cancel,
+		client: client,
+		stream: stream,
+		name:   name,
+	}
+}
+
+// Name returns the operation name, if known.
+func (s *RetryingStream) Name() string {
+	return s.name
+}
+
+// Recv attempts to reliably return the next operation on the named stream.
+//
+// If the stream is disconnected and the operation name has been received, it
+// will attempt to reconnect with WaitExecution.
+func (s *RetryingStream) Recv() (*longrunning.Operation, error) {
+	r := retry.DefaultWithContext(s.ctx)
+	for {
+		op, err := s.stream.Recv()
+		if err == nil {
+			if op.GetName() != "" {
+				s.name = op.GetName()
+			}
+			return op, nil
+		}
+		if !status.IsUnavailableError(err) || s.name == "" {
+			return nil, err
+		}
+		if !r.Next() {
+			return nil, s.ctx.Err()
+		}
+		req := &repb.WaitExecutionRequest{Name: s.name}
+		next, err := s.client.WaitExecution(s.ctx, req)
+		if err != nil {
+			return nil, err
+		}
+		s.stream.CloseSend()
+		s.stream = next
+	}
+}
+
+func (s *RetryingStream) CloseSend() error {
+	var err error
+	if s.stream != nil {
+		err = s.stream.CloseSend()
+		s.stream = nil
+	}
+	s.client = nil
+	s.cancel()
+	return err
+}
+
+// ExecuteOperation contains an operation along with its execution-specific
+// payload.
+type ExecuteOperation struct {
```

Comment:
private?

Reply:
after refactoring, Wait() returns one of these, so kept it public.